### PR TITLE
feat(ec2): support throughput on LaunchTemplate EBS volumes

### LINKED ISSUE #24341: EbsDeviceProps missing support for setting throughput on GP3 volumes, yet VolumeProps has support for it
### Describe the feature

When creating and EC2 instance, and configuring a blockDevices property, this uses the [BlockDeviceVolume](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ec2.BlockDeviceVolume.html) type, which accepts an ebsDevice parameter of type EbsDeviceProps. EbsDeviceProps does not have the option to set a throughput property.

However, if you look at [VolumeProps](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ec2.VolumeProps.html), when creating a volume separately, it does support the throughput property. 

### Use Case

I've seen examples where people are creating volumes outside of the blockDevices property of an Instance, then have to deal with attaching the volume. 

Having the option to create a gp3 volume along with all of its configuration properties (including throughput) directly via the instance as part of the blockDevices property simplifies the deployment of an instance. iops is already an accepted property as part of EbsDeviceProps, why not throughput?

### Proposed Solution

_No response_

### Other Information

_No response_

### Acknowledgements

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.66.1

### Environment details (OS name and version, etc.)

OSX 12.6.3

### PR DESCRIPTION
**Pull request by isker**
### Issue # (if applicable)

Fixes #24341.

### Reason for this change

You currently can't specify `throughput` on LaunchTemplate EBS volumes.

### Description of changes

This support was simply not included in ec2 when it was added to autoscaling in #22441. I have copied that PR's implementation implementation to ec2 and similarly adapted its tests.

### Description of how you validated changes

Unit and integration tests.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by isker** [comment]
The build failure:
```
[4/4] Building fresh packages...
error /codebuild/output/src3794156442/src/github.com/aws/aws-cdk/node_modules/@lerna/create/node_modules/nx, /codebuild/output/src3794156442/src/github.com/aws/aws-cdk/node_modules/@nx/devkit/node_modules/nx, /codebuild/output/src3794156442/src/github.com/aws/aws-cdk/node_modules/lerna/node_modules/nx: Command failed.
Exit code: 135
Command: node ./bin/post-install
Arguments: 
Directory: /codebuild/output/src3794156442/src/github.com/aws/aws-cdk/node_modules/lerna/node_modules/nx
Output:
Bus error (core dumped)
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

[Container] 2024/07/01 02:44:29.448680 Command did not exit successfully /bin/bash ./build.sh --ci exit status 135
[Container] 2024/07/01 02:44:29.451865 Phase complete: BUILD State: FAILED
```

This is nowhere near the first spurious build failure I've seen in just my third PR to this project, and the problems are not limited to this main CI job: https://github.com/aws/aws-cdk/actions/runs/9737293345/job/26869186116?pr=30716 has also failed for reasons unclear.

It has proven to be challenging to contribute to aws-cdk. Contributors need to pass through an underspecified battery of PR label transitions in order that their PR be brought to a human being's attention so that it may be reviewed and merged. These transitions are automated and gated by CI success. But here is a PR where two of the gating CI jobs have spuriously failed, with no way for me to initiate a retry except by, I guess, rewording my commit message so that I get a different commit hash. 

And if contributors don't teach themselves to do this song and dance then their PR will get closed by automation after a period of inactivity, as happened to this changeset in #30317.

Suggestions to maintainers:
1. Please specify the actual process for PR review so that contributors understand what is expected of them. Having to spend substantial time trying to infer the process based upon my interactions with flakey CI jobs and occasionally maintainers is discouraging. https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md#getting-a-review-from-a-maintainer does not specify:
  a. What to do if you can't get this project's flakey CI jobs to pass.
  b. What an exemption request is.
  c. What to do if your exemption requests are being ignored. I ended up having to repeatedly ping maintainers I thought could help me on #30317, to no avail. I hope this process was nowhere near as embarrassing to them as it was to me.
2. If you cannot produce reliable CI, please stop gating PR review progression on it.
3. The automated "close stale PRs" job leaves something to be desired. On #30317, this job closed the PR. I successfully begged a maintainer to reopen it. The job then quickly closed it again. (I gave up for a week, and am now back with a new PR.) Having been reopened by a maintainer, the job should not have quickly closed the PR again.

In aggregate, you have crafted a fully automated, drawn out process for contributors to be ignored. Ignoring contributors is your prerogative as maintainers, but you could be achieving such outcomes with much less effort than this.

**Comment by isker** [comment]
Screenshot 2024-06-30 at 23 42 01

Well, thanks, I guess.

File: packages/aws-cdk-lib/aws-ec2/lib/volume.ts [hunk 1]
```diff
@@ -70,6 +70,14 @@ export interface EbsDeviceOptionsBase {
    * `@aws-cdk/aws-ec2:ebsDefaultGp3Volume` is enabled.
    */
   readonly volumeType?: EbsDeviceVolumeType;
+
+  /**
+   * The throughput that the volume supports, in MiB/s
+   * Takes a minimum of 125 and maximum of 1000.
+   * @see https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSVolumeTypes.html
+   * @default - 125 MiB/s. Only valid on gp3 volumes.
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-ec2/lib/volume.ts:78]:
```suggestion
   * The throughput to provision for a `gp3` volume.
   *
   * Valid Range: Minimum value of 125. Maximum value of 1000.
   *
   * @see https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSVolumeTypes.html
   *
   * @default - 125 MiB/s.
```

File: packages/aws-cdk-lib/aws-ec2/lib/private/ebs-util.ts [hunk 2]
```diff
@@ -25,7 +25,30 @@ function synthesizeBlockDeviceMappings<RT, NDT>(construct: Construct, blockDevic
...
+
+        if (volumeType != EbsDeviceVolumeType.GP3) {
+          throw new Error('throughput property requires volumeType: EbsDeviceVolumeType.GP3');
+        }
+
+        if (throughput < Min || throughput > Max) {
+          throw new Error(
+            `throughput property takes a minimum of ${Min} and a maximum of ${Max}`,
+          );
+        }
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-ec2/lib/private/ebs-util.ts:42]:
```suggestion
        if (volumeType != EbsDeviceVolumeType.GP3) {
          throw new Error(`throughput requires volumeType: ${EbsDeviceVolumeType.GP3}, got: ${volumeType}.`);
        }

        if (throughput < 125 || throughput > 1000) {
          throw new Error(`throughput must be between 125 and 1000, got ${throughput}.`);
        }
```
Let's keep it cleaner and provide more feedback on the error message.

File: packages/aws-cdk-lib/aws-ec2/lib/private/ebs-util.ts [hunk 3]
```diff
@@ -25,7 +25,30 @@ function synthesizeBlockDeviceMappings<RT, NDT>(construct: Construct, blockDevic
...
+
+        const maximumThroughputRatio = 0.25;
+        if (iops) {
+          const iopsRatio = (throughput / iops);
+          if (iopsRatio > maximumThroughputRatio) {
+            throw new Error(`Throughput (MiBps) to iops ratio of ${iopsRatio} is too high; maximum is ${maximumThroughputRatio} MiBps per iops`);
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-ec2/lib/private/ebs-util.ts:48]: Can you provide documentation about this rule?

File: packages/aws-cdk-lib/aws-ec2/README.md [hunk 4]
```diff
@@ -1543,8 +1543,8 @@ new ec2.Instance(this, 'Instance', {
 ### Block Devices
 
 To add EBS block device mappings, specify the `blockDevices` property. The following example sets the EBS-backed
-root device (`/dev/sda1`) size to 50 GiB, and adds another EBS-backed device mapped to `/dev/sdm` that is 100 GiB in
-size:
+root device (`/dev/sda1`) size to 50 GiB, and adds another EBS-backed GP3 device mapped to `/dev/sdm` that is 100 GiB in
+size, with the GP3-specific `throughput` option set to 250 MiB/s:
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-ec2/README.md:1547]: Can you please add a separate paragraph for the new prop at the end of the `Block Devices` section? Eg
```
To specify the throughput value for `gp3` volumes, use the `throughput` property:

```ts
declare const vpc: ec2.Vpc;
declare const instanceType: ec2.InstanceType;
declare const machineImage: ec2.IMachineImage;

const kmsKey = new Key(this, 'KmsKey')

new ec2.Instance(this, 'Instance', {
  vpc,
  instanceType,
  machineImage,

  // ...

  blockDevices: [
    {
      deviceName: '/dev/sda1',
      volume: ec2.BlockDeviceVolume.ebs(100, {
        volumeType: ec2.EbsDeviceVolumeType.GP3,
        throughput: 250,
      }),
    },
  ],
});

**Review by lpizzinidev — CHANGES_REQUESTED** [review]
Thanks 👍 Left some comments for adjustments

File: packages/aws-cdk-lib/aws-ec2/lib/volume.ts (hunk 1 above)
Comment (isker) [reply packages/aws-cdk-lib/aws-ec2/lib/volume.ts:78]: For this and for your other code feedback below, I have copied this verbatim from the existing code for autoscaling. For example: https://github.com/aws/aws-cdk/blob/da2ec75f6bc8f2e28b07cdf1307b22c83bb652d5/packages/aws-cdk-lib/aws-autoscaling/lib/auto-scaling-group.ts#L2430-L2451

Should we keep these consistent or?

Comment (lpizzinidev) [reply packages/aws-cdk-lib/aws-ec2/lib/volume.ts:78]: I'd go for the cleaner version and provide some documentation about the `0.25` ratio if possible, thanks.

File: packages/aws-cdk-lib/aws-ec2/lib/private/ebs-util.ts (hunk 3 above)
Comment (isker) [reply packages/aws-cdk-lib/aws-ec2/lib/private/ebs-util.ts:48]: I cargo culted it from https://github.com/aws/aws-cdk/pull/22441, but after some struggle I did manage to find the documentation that this check is seemingly based on [here](https://docs.aws.amazon.com/ebs/latest/userguide/general-purpose.html#gp3-performance).
> gp3 volumes deliver a consistent baseline throughput performance of 125 MiB/s, which is included with the price of storage. You can provision additional throughput (up to a maximum of 1,000 MiB/s) for an additional cost at a ratio of 0.25 MiB/s per provisioned IOPS. Maximum throughput can be provisioned at 4,000 IOPS or higher and 8 GiB or larger (4,000 IOPS × 0.25 MiB/s per IOPS = 1,000 MiB/s).

Do you want this documentation mentioned somewhere in code?

**Comment by isker** [comment]
https://github.com/aws/aws-cdk/pull/30716#issuecomment-2259539502 the log file is completely empty?

File: packages/aws-cdk-lib/aws-ec2/lib/volume.ts [hunk 5]
```diff
@@ -70,6 +70,17 @@ export interface EbsDeviceOptionsBase {
    * `@aws-cdk/aws-ec2:ebsDefaultGp3Volume` is enabled.
    */
   readonly volumeType?: EbsDeviceVolumeType;
+
+  /**
+   * The throughput to provision for a `gp3` volume.
+   *
+   * Valid Range: Minimum value of 125. Maximum value of 1000.
+   *
+   * @see https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSVolumeTypes.html
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-ec2/lib/volume.ts:79]:
```suggestion
   * Valid Range: Minimum value of 125. Maximum value of 1000.
   *
   * `gp3` volumes deliver a consistent baseline throughput performance of 125 MiB/s.
   * You can provision additional throughput for an additional cost at a ratio of 0.25 MiB/s per provisioned IOPS.
   *
   * @see https://docs.aws.amazon.com/ebs/latest/userguide/general-purpose.html#gp3-performance
```

File: packages/aws-cdk-lib/aws-ec2/lib/private/ebs-util.ts (hunk 3 above)
Comment (lpizzinidev) [reply packages/aws-cdk-lib/aws-ec2/lib/private/ebs-util.ts:43]: Thanks 👍 Yes I'd update the comment to the variable to specify the behavior (see other comment).

**Review by lpizzinidev — APPROVED** [review]
Thanks 👍

File: packages/aws-cdk-lib/aws-ec2/lib/private/ebs-util.ts [hunk 6]
```diff
@@ -25,7 +25,25 @@ function synthesizeBlockDeviceMappings<RT, NDT>(construct: Construct, blockDevic
...
-      const { iops, volumeType, kmsKey, ...rest } = ebs;
+      const { iops, throughput, volumeType, kmsKey, ...rest } = ebs;
+
+      if (throughput) {
+        if (volumeType != EbsDeviceVolumeType.GP3) {
+          throw new Error(`throughput requires volumeType: ${EbsDeviceVolumeType.GP3}, got: ${volumeType}.`);
+        }
+
+        if (throughput < 125 || throughput > 1000) {
+          throw new Error(`throughput must be between 125 and 1000, got ${throughput}.`);
```
Comment (GavinZZ) [inline packages/aws-cdk-lib/aws-ec2/lib/private/ebs-util.ts:36]:
```suggestion
          throw new Error(`'throughput' must be between 125 and 1000, got ${throughput}.`);
```

**Review by GavinZZ — DISMISSED** [review]
Maybe I missed it, but I only saw the validation and definition of the property `throughput`, but I don't see where you assign it to `CfnVolume`?

File: packages/aws-cdk-lib/aws-ec2/lib/volume.ts [hunk 7]
```diff
@@ -70,6 +70,20 @@ export interface EbsDeviceOptionsBase {
...
+   * Valid Range: Minimum value of 125. Maximum value of 1000.
+   *
+   * `gp3` volumes deliver a consistent baseline throughput performance of 125 MiB/s.
+   * You can provision additional throughput for an additional cost at a ratio of 0.25 MiB/s per provisioned IOPS.
+   *
+   * @see https://docs.aws.amazon.com/ebs/latest/userguide/general-purpose.html#gp3-performance
+   *
+   * @default - 125 MiB/s.
+   */
+  readonly throughput?: number;
```
[the remark this answers is not in the document]
Comment (isker) [reply packages/aws-cdk-lib/aws-ec2/lib/volume.ts:86]: [No](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-ec2-volume.html#cfn-ec2-volume-throughput). Is there a better type to use?

**Comment by isker** [comment]
> Maybe I missed it, but I only saw the validation and definition of the property `throughput`, but I don't see where you assign it to `CfnVolume`?

https://github.com/aws/aws-cdk/pull/30716/files#diff-c1fd5c618720a81ee00a0360ca37364d5d5281c596a5bc8f63edb637607b6d54R64

This is not using `CfnVolume`. It's using a (nested) property of LaunchTemplate: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-ebs.html.

**Comment by GavinZZ** [comment]
> > Maybe I missed it, but I only saw the validation and definition of the property `throughput`, but I don't see where you assign it to `CfnVolume`?
> 
> https://github.com/aws/aws-cdk/pull/30716/files#diff-c1fd5c618720a81ee00a0360ca37364d5d5281c596a5bc8f63edb637607b6d54R64
> 
> This is not using `CfnVolume`. It's using a (nested) property of LaunchTemplate: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-launchtemplate-ebs.html https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_ec2.CfnLaunchTemplate.EbsProperty.html#throughput.

Ah I see, I missed that line of code. Thanks for the clarification.

File: packages/aws-cdk-lib/aws-ec2/lib/volume.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (GavinZZ) [reply packages/aws-cdk-lib/aws-ec2/lib/volume.ts:86]: `number` is fine, but we can do a validation something like `Number.isInteger` when you do the validation above.

**Review by GavinZZ — APPROVED** [review]
LGTM, thank you!

**Pull request merged**